cnd/src/swap_protocols/herc20.rs

```diff
@@ -0,0 +1,263 @@
+
+#[async_trait::async_trait]
+impl state::Update<Event> for StateStore {
+    async fn update(&self, key: &SwapId, event: Event) {
+        let mut states = self.states.lock().await;
+        let entry = states.entry(*key);
+
+        match (event, entry) {
+            (Event::Started, Entry::Vacant(vacant)) => {
+                vacant.insert(State::None);
+            }
+            (Event::Deployed(deployed), Entry::Occupied(mut state)) => {
+                state.get_mut().transition_to_deployed(deployed)
+            }
+            (Event::Funded(funded), Entry::Occupied(mut state)) => {
+                state.get_mut().transition_to_funded(funded)
+            }
+            (Event::Redeemed(redeemed), Entry::Occupied(mut state)) => {
+                state.get_mut().transition_to_redeemed(redeemed)
+            }
+            (Event::Refunded(refunded), Entry::Occupied(mut state)) => {
+                state.get_mut().transition_to_refunded(refunded)
+            }
+            (Event::Started, Entry::Occupied(_)) => {
+                tracing::warn!(
+                    "Received Started event for {} although state is already present",
+                    key
+                );
+            }
+            (_, Entry::Vacant(_)) => {
+                tracing::warn!("State not found for {}", key);
+            }
+        }
+    }
+}
+
+/// Creates a new instance of the herc20 protocol.
+///
+/// Returns a stream of events happening during the execution.
+pub fn new<'a, C>(
+    connector: &'a C,
+    params: Params,
+) -> impl Stream<Item = anyhow::Result<Event>> + 'a
+where
+    C: Deployed + Funded + Redeemed + Refunded,
+{
+    Gen::new({
+        |co| async move {
+            if let Err(error) = watch_ledger(connector, params, &co).await {
+                co.yield_(Err(error)).await;
+            }
+        }
+    })
+}
+
+async fn watch_ledger<C, R>(
+    connector: &C,
+    params: Params,
+    co: &Co<anyhow::Result<Event>, R>,
+) -> anyhow::Result<()>
+where
+    C: Deployed + Funded + Redeemed + Refunded,
+{
+    co.yield_(Ok(Event::Started)).await;
+
+    let deployed = connector.deployed(params.clone()).await?;
+
+    co.yield_(Ok(Event::Deployed(deployed.clone()))).await;
+
+    let funded = connector.funded(params.clone(), deployed.clone()).await?;
+    co.yield_(Ok(Event::Funded(funded))).await;
+
+    let redeemed = connector.redeemed(params.clone(), deployed.clone());
+    let refunded = connector.refunded(params, deployed);
+
+    match future::try_select(redeemed, refunded).await {
+        Ok(Either::Left((redeemed, _))) => {
+            co.yield_(Ok(Event::Redeemed(redeemed))).await;
+        }
+        Ok(Either::Right((refunded, _))) => {
+            co.yield_(Ok(Event::Refunded(refunded))).await;
+        }
+        Err(either) => {
+            let (error, _other_future) = either.factor_first();
+            return Err(error);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Clone, Debug)]
+pub struct Params {
```

Comment:
These parameters are not orthogonal to each other.

Reply:
The asset is inside the htlc and so is the secret hash. I would say just compute the HTLC from the remaining parameters like we used to.

Reply:
Ah yes indeed, apologies on this one, I intended to clean it up once it compiled but did not give myself a todo.